modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinStatement.java

```diff
@@ -228,57 +230,79 @@ protected void execute0(JdbcStatementType stmtType, String sql, List<Object> arg
+            if (res0 instanceof JdbcQueryExecuteResult) {
+                JdbcQueryExecuteResult res = (JdbcQueryExecuteResult)res0;
 
-            resultSets = new ArrayList<>(resInfos.size());
+                resultSets = Collections.singletonList(new JdbcThinResultSet(this, res.cursorId(), pageSize,
+                    res.last(), res.items(), res.isQuery(), conn.autoCloseServerCursor(), res.updateCount(),
+                    closeOnCompletion, stickyIo));
+            }
+            else if (res0 instanceof JdbcQueryExecuteMultipleStatementsResult) {
+                JdbcQueryExecuteMultipleStatementsResult res = (JdbcQueryExecuteMultipleStatementsResult)res0;
 
-            boolean firstRes = true;
+                List<JdbcResultInfo> resInfos = res.results();
 
-            for (JdbcResultInfo rsInfo : resInfos) {
-                if (!rsInfo.isQuery())
-                    resultSets.add(resultSetForUpdate(rsInfo.updateCount()));
-                else {
-                    if (firstRes) {
-                        firstRes = false;
+                resultSets = new ArrayList<>(resInfos.size());
 
-                        resultSets.add(new JdbcThinResultSet(this, rsInfo.cursorId(), pageSize, res.isLast(),
-                            res.items(), true, conn.autoCloseServerCursor(), -1, closeOnCompletion,
-                            stickyIo));
-                    }
+                boolean firstRes = true;
+
+                for (JdbcResultInfo rsInfo : resInfos) {
+                    if (!rsInfo.isQuery())
+                        resultSets.add(resultSetForUpdate(rsInfo.updateCount()));
                     else {
-                        resultSets.add(new JdbcThinResultSet(this, rsInfo.cursorId(), pageSize, false,
-                            null, true, conn.autoCloseServerCursor(), -1, closeOnCompletion,
-                            stickyIo));
+                        if (firstRes) {
+                            firstRes = false;
+
+                            resultSets.add(new JdbcThinResultSet(this, rsInfo.cursorId(), pageSize, res.isLast(),
+                                res.items(), true, conn.autoCloseServerCursor(), -1, closeOnCompletion,
+                                stickyIo));
+                        }
+                        else {
+                            resultSets.add(new JdbcThinResultSet(this, rsInfo.cursorId(), pageSize, false,
+                                null, true, conn.autoCloseServerCursor(), -1, closeOnCompletion,
+                                stickyIo));
+                        }
                     }
                 }
             }
+            else
+                throw new SQLException("Unexpected result [res=" + res0 + ']');
+
+            if (autoCommit && txEnabled)
+                txCtx.end(true);
+        }
+        catch (Exception e) {
+            // Rollback in case of error.
+            if (autoCommit && txEnabled)
+                txCtx.end(false);
```

Comment:
Shouldn't we clear JdbcThinConnection.txCtx?

Reply:
It will be null here - see JdbcThinConnection#transactionContext implementation
```java
    TxContext transactionContext() throws SQLException {
...
        TxContext txCtx0 = new TxContext(txIsolation, txConcurrency);

        if (!autoCommit)
            txCtx = txCtx0;

        return txCtx0;
    }
```